services/venice-controller/src/main/java/com/linkedin/venice/pushmonitor/PushStatusCollector.java

```diff
@@ -0,0 +1,231 @@
+  }
+
+  private void scanDaVinciPushStatus() {
+    List<CompletableFuture<TopicPushStatus>> resultList = new ArrayList<>();
+    for (Map.Entry<String, TopicPushStatus> entry: topicToPushStatusMap.entrySet()) {
+      String topicName = entry.getKey();
+      TopicPushStatus pushStatus = entry.getValue();
+      if (!pushStatus.isMonitoring()) {
+        continue;
+      }
+      if (pushStatus.getDaVinciStatus() != null && pushStatus.getDaVinciStatus().getStatus().isTerminal()) {
+        resultList.add(CompletableFuture.completedFuture(pushStatus));
+      } else {
+        resultList.add(CompletableFuture.supplyAsync(() -> {
+          ExecutionStatusWithDetails statusWithDetails = PushMonitorUtils.getDaVinciPushStatusAndDetails(
+              pushStatusStoreReader,
+              topicName,
+              pushStatus.getPartitionCount(),
+              Optional.empty());
+          pushStatus.setDaVinciStatus(statusWithDetails);
+          return pushStatus;
+        }, pushStatusStoreScanExecutor));
+      }
+    }
+    // Collect the executor result and compute aggregate results for ongoing pushes.
+    for (CompletableFuture<TopicPushStatus> future: resultList) {
+      TopicPushStatus pushStatus;
+      try {
+        pushStatus = future.get();
+      } catch (Exception e) {
+        LOGGER.error("Caught exception when getting future result of push status.", e);
+        continue;
+      }
+      ExecutionStatusWithDetails daVinciStatus = pushStatus.getDaVinciStatus();
+      ExecutionStatusWithDetails serverStatus = pushStatus.getServerStatus();
+      if (serverStatus == null) {
+        continue;
+      }
+      LOGGER.info(
+          "Topic server push status: {}, Da Vinci push status: {}",
+          serverStatus.getStatus(),
+          daVinciStatus.getStatus());
+      if (isOverallPushCompleted(serverStatus, daVinciStatus)) {
+        pushStatus.setMonitoring(false);
+        pushCompletedHandler.accept(pushStatus.getTopicName());
+      } else if (isOverallPushError(serverStatus, daVinciStatus)) {
+        pushStatus.setMonitoring(false);
+        StringBuilder pushErrorDetailStringBuilder = new StringBuilder();
+        if (serverStatus.getStatus().equals(ExecutionStatus.ERROR)) {
+          pushErrorDetailStringBuilder.append("Server push error: ").append(serverStatus.getDetails()).append("\n");
+        }
+        if (daVinciStatus.getStatus().equals(ExecutionStatus.ERROR)) {
+          pushErrorDetailStringBuilder.append("Da Vinci push error: ").append(daVinciStatus.getDetails()).append("\n");
+        }
+        pushErrorHandler.accept(pushStatus.getTopicName(), pushErrorDetailStringBuilder.toString());
+      }
+    }
+  }
+
+  public void handleServerPushStatusUpdate(String topicName, ExecutionStatus executionStatus, String detailsString) {
+
+    TopicPushStatus topicPushStatus = topicToPushStatusMap.get(topicName);
+    if ((!daVinciPushStatusScanEnabled) || topicPushStatus == null) {
```

Comment:
can this condition happen when `daVinciPushStatusScanEnabled` is `true` ?
```
topicPushStatus == null
```

Reply:
It seems when the push status system store is not enabled in store-level, this condition can be true.

Reply:
Can you add some javadoc for the above scenario?

Reply:
Actually this is done in the latest commit.